olive/passes/onnx/genai_model_exporter.py

```diff
@@ -0,0 +1,76 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# Export a PyTorch model using the onnxruntime-genai package.
+# --------------------------------------------------------------------------
+import logging
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict
+
+from onnx import TensorProto
+
+from olive.hardware.accelerator import AcceleratorLookup, AcceleratorSpec, Device
+from olive.model import ONNXModelHandler, PyTorchModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.olive_pass import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class GenAIModelExporter(Pass):
+    class Precision(str, Enum):
+        FP32 = "fp32"
+        FP16 = "fp16"
+        INT4 = "int4"
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "precision": PassConfigParam(
+                type_=GenAIModelExporter.Precision,
+                required=True,
+                description="Precision of model.",
+            )
+        }
+
+    def _run_for_config(
+        self, model: PyTorchModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModelHandler:
+        from onnxruntime_genai.models.builder import create_model
+
+        Path(output_model_path).mkdir(parents=True, exist_ok=True)
+        output_model_filepath = Path(resolve_onnx_path(output_model_path))
+
+        precision = config["precision"]
+        device = (
+            Device.CPU
+            if self.accelerator_spec.execution_provider
+            in AcceleratorLookup.get_execution_providers_for_device(Device.CPU)
+            else Device.GPU
+        )
+
+        logger.info(
+            "Valid precision + execution provider combinations are: FP32 CPU, FP32 CUDA, FP16 CUDA, INT4 CPU, INT4 CUDA"
+        )
+
+        # Set input/output precision of ONNX model
+        io_dtype = (
+            TensorProto.FLOAT
+            if precision in {"int8", "fp32"} or (precision == "int4" and device == Device.CPU)
+            else TensorProto.FLOAT16
+        )
+
+        create_model(
+            model_name_or_path=str(model.hf_config.model_name),
+            output_dir=str(output_model_filepath.parent),
+            precision=str(precision),
+            execution_provider=str(device),
+            cache_dir=str(Path.cwd() / "cache_dir"),
```

Comment:
Is that possible to update the pass internal cache_dir as `output_model_filepath.parent/"cache_dir"`? To ensure the data generated by pass is all under the corresponding cache folder.

Reply:
I intentionally kept the GenAI cache out of Olive cache to avoid having to download the model over & over again. During development Olive cache gets deleted which will force a download again. Ideally, GenAI should just use the HF cache and not bother with the separate script specific cache at all but unfortunately, that's not in our control.

Reply:
maybe we can use a cache name like `genai_cache` or something since `cache_dir` doesn't indicate what the cache is for and the source is hidden from the user. 
And also add `genai_cache/` to the repository's root `.gitignore`

Reply:
```
        cache_dir = os.environ.get("HF_HOME", None)
        if not cache_dir:
            cache_dir = os.environ.get("TRANSFORMERS_CACHE", None)
        if not cache_dir:
            cache_dir = output_model_filepath.parent / "genai_cache_dir"
```